core/src/main/java/io/questdb/griffin/engine/functions/date/DatediffTimestampFunctionFactory.java

```diff
@@ -0,0 +1,567 @@
+/*******************************************************************************
+ *     ___                  _   ____  ____
+ *    / _ \ _   _  ___  ___| |_|  _ \| __ )
+ *   | | | | | | |/ _ \/ __| __| | | |  _ \
+ *   | |_| | |_| |  __/\__ \ |_| |_| | |_) |
+ *    \__\_\\__,_|\___||___/\__|____/|____/
+ *
+ *  Copyright (c) 2014-2019 Appsicle
+ *  Copyright (c) 2019-2020 QuestDB
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.griffin.engine.functions.date;
+
+import io.questdb.cairo.CairoConfiguration;
+import io.questdb.cairo.sql.Function;
+import io.questdb.cairo.sql.Record;
+import io.questdb.griffin.FunctionFactory;
+import io.questdb.griffin.engine.functions.LongFunction;
+import io.questdb.griffin.engine.functions.UnaryFunction;
+import io.questdb.griffin.engine.functions.BinaryFunction;
+import io.questdb.griffin.engine.functions.TernaryFunction;
+import io.questdb.griffin.engine.functions.constants.TimestampConstant;
+import io.questdb.std.Numbers;
+import io.questdb.std.ObjList;
+import io.questdb.std.microtime.Timestamps;
+
+public class DatediffTimestampFunctionFactory implements FunctionFactory {
+    @Override
+    public String getSignature() {
+        return "datediff(ANN)";
+    }
+
+    @Override
+    public Function newInstance(ObjList<Function> args, int position, CairoConfiguration configuration) {
+
+        Function period = args.getQuick(0);
+        if (period.isConstant()) {
+            Function start = args.getQuick(1);
+            Function end = args.getQuick(2);
+            char periodValue = period.getChar(null);
+            if (periodValue == 's') {
+                if (start.isConstant() && start.getTimestamp(null) != Numbers.LONG_NaN) {
+                    long startValue = start.getLong(null);
+                    return new DateDiffFuncSecondConstantTimestampConstant(position, args.getQuick(2), startValue);
+                }
+                if (end.isConstant() && end.getTimestamp(null) != Numbers.LONG_NaN) {
+                    long endValue = end.getLong(null);
+                    return new DateDiffFuncSecondConstantTimestampConstant(position, args.getQuick(1), endValue);
+                }
+                return new DateDiffFuncSecondConstant(position, args.getQuick(1), args.getQuick(2));
+            }
+            if (periodValue == 'm') {
```

Comment:
there is a lot of boilerplate here.

if you introduce an interface, such as

```java
    @FunctionalInterface
    private interface DiffConstructorRightIsConst {
        LongFunction build(int position, Function left, long right);
    }
```

and have them mapped to letters in an array then you can write all the const & null logic once